op-batcher/batcher/espresso/streamer.go

```diff
@@ -0,0 +1,232 @@
+	// Namespace of the rollup we're interested in
+	Namespace uint64
+	// Address of the batcher, we expect transactions to
+	// be signed by the corresponding private key
+	BatcherAddress common.Address
+
+	L1Client            L1Client
+	EspressoClient      *espressoClient.Client
+	EspressoLightClient *espressoLightClient.LightClientReader
+	Log                 log.Logger
+
+	// Batch number we're to give out next
+	BatchPos uint64
+	// HotShot block we're to fetch next
+	hotShotPos uint64
+	// Position of the last safe batch
+	confirmedBatchPos uint64
+	// Hotshot block corresponding to the last safe batch
+	confirmedHotShotPos uint64
+
+	// Maintained in sorted order, but may be missing batches if we receive
+	// any out of order.
+	batchBuffer []EspressoBatch
+}
+
+// Reset the state to the last safe batch
+func (s *EspressoStreamer) Reset() {
+	s.BatchPos = s.confirmedBatchPos + 1
+	s.hotShotPos = s.confirmedHotShotPos
+	s.batchBuffer = nil
+}
+
+// Handle both L1 reorgs and batcher restarts by updating our state in case it is
+// not consistent with what's on the L1. Returns true if the state was updated.
+func (s *EspressoStreamer) Refresh(ctx context.Context, syncStatus *eth.SyncStatus) (bool, error) {
+	if s.confirmedBatchPos == syncStatus.SafeL2.Number {
+		return false, nil
+	}
+
+	hotshotState, err := s.EspressoLightClient.LightClient.
+		FinalizedState(&bind.CallOpts{BlockNumber: new(big.Int).SetUint64(syncStatus.SafeL2.L1Origin.Number)})
+	if err != nil {
+		return false, err
+	}
+
+	s.confirmedBatchPos = syncStatus.SafeL2.Number
+	s.confirmedHotShotPos = hotshotState.BlockHeight
+	s.Reset()
+	return true, nil
+}
+
+func (s *EspressoStreamer) Update(ctx context.Context) error {
+	// Fetch more batches from HotShot if available.
+	hotshotState, err := s.EspressoLightClient.LightClient.FinalizedState(&bind.CallOpts{})
+
+	if err != nil {
+		return fmt.Errorf("failed to fetch HotShot block height: %w", err)
+	}
+
+	s.Log.Debug("Updated finalized hotshot state", "hotshotState", hotshotState)
+
+	targetHeight := min(hotshotState.BlockHeight, s.hotShotPos+100)
+
+	for ; s.hotShotPos < targetHeight; s.hotShotPos += 1 {
+		s.Log.Debug("fetching HotShot block", "blockNr", s.hotShotPos)
+
+		txns, err := s.EspressoClient.FetchTransactionsInBlock(ctx, s.hotShotPos, s.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to fetch transactions in block: %w", err)
+		}
+
+		if len(txns.Transactions) == 0 {
+			s.Log.Debug("no transactions in hotshot block", "blockNr", s.hotShotPos)
+			continue
+		}
+
+		rawHeader, err := s.EspressoClient.FetchRawHeaderByHeight(ctx, s.hotShotPos)
+		if err != nil {
+			return fmt.Errorf("failed to fetch raw HotShot header: %w", err)
+		}
+
+		var header espressoTypes.HeaderImpl
+		err = json.Unmarshal(rawHeader, &header)
+		if err != nil {
+			return fmt.Errorf("could not unmarshal header from bytes")
+		}
+
+		snapshot, err := s.EspressoLightClient.FetchMerkleRoot(s.hotShotPos, nil)
```

Comment:
As noted by Jeb above, this logic is meant to be replaced by querying various Espresso nodes right?